dataplane/src/nat.rs

```diff
@@ -19,88 +27,60 @@ struct Pif {
-struct PrefixTrie<K, V>
-where
-    K: Clone + Eq + Hash,
-    V: Clone,
-{
-    root: TrieNode<K, V>,
+#[derive(Default, Clone)]
+struct PrefixTrie {
+    trie_ipv4: RTrieMap<Ipv4Prefix, String>,
+    trie_ipv6: RTrieMap<Ipv6Prefix, String>,
 }
 
-impl<K, V> PrefixTrie<K, V>
-where
-    K: Clone + Eq + Hash + Default + Debug,
-    V: Clone + Default + Debug,
-{
+impl PrefixTrie {
     #[tracing::instrument(level = "trace")]
     fn new() -> Self {
         Self {
-            root: TrieNode::default(),
+            trie_ipv4: RTrieMap::new(),
+            trie_ipv6: RTrieMap::new(),
         }
     }
 
     #[tracing::instrument(level = "trace")]
-    fn insert(&mut self, keys: Vec<K>, value: V) -> bool {
-        let mut node = &mut self.root;
-
-        for key in keys {
-            node = node.children.entry(key).or_default();
-        }
-
-        if node.value.is_some() {
-            return false; // Prefix already exists
-        }
-
-        node.value = Some(value);
-        true
-    }
-
-    #[tracing::instrument(level = "trace")]
-    fn find(&self, keys: Vec<K>) -> Option<V> {
-        let mut node = &self.root;
-        let mut best_match = None;
-
-        for key in keys {
-            if let Some(val) = &node.value {
-                best_match = Some(val.clone());
+    fn insert(&mut self, addr: &IpAddr, len: &u8, value: String) -> Result<(), NatError> {
+        match addr {
+            IpAddr::V4(ip) => {
+                let prefix = Ipv4Prefix::new(*ip, *len).or(Err(NatError::BadPrefix))?;
+                if self.trie_ipv4.insert(prefix, value).is_some() {
+                    return Err(NatError::PifExists);
+                }
             }
-            if let Some(child) = node.children.get(&key) {
-                node = child;
-            } else {
-                break;
+            IpAddr::V6(ip) => {
+                let prefix = Ipv6Prefix::new(*ip, *len).or(Err(NatError::BadPrefix))?;
+                if self.trie_ipv6.insert(prefix, value).is_some() {
+                    return Err(NatError::PifExists);
+                }
             }
         }
-
-        best_match
+        Ok(())
     }
 
     #[tracing::instrument(level = "trace")]
-    fn ip_to_bits(ip: &IpAddr, prefix_len: u8) -> Vec<u8> {
-        let mut bits = Vec::new();
-        match ip {
-            IpAddr::V4(ipv4) => {
-                for i in 0..prefix_len {
-                    bits.push((ipv4.octets()[i as usize / 8] >> (7 - (i % 8))) & 1);
-                }
-            }
-            IpAddr::V6(ipv6) => {
-                for i in 0..prefix_len {
-                    bits.push((ipv6.octets()[i as usize / 8] >> (7 - (i % 8))) & 1);
-                }
-            }
+    fn find(&self, ip: &IpAddr) -> Option<String> {
+        let res = match ip {
+            IpAddr::V4(ip) => self.trie_ipv4.lookup(ip).1,
+            IpAddr::V6(ip) => self.trie_ipv6.lookup(ip).1,
+        };
+        if res.is_empty() {
```

Comment:
Here you're using the fact that the RTrieMap always has a root object, correct?
When you will add a delete() method, be careful to forbid deletions for 0/0 and 0::/0.
Else the RtrieMap will panic.

Reply:
Yes, the lookup always returns something, so I've got to ensure the value we get is different from the one attached to the root - which in our case is the default value for `String`, an empty string.

I could probably add a comment to make it clear.

I'll try to keep that in mind for the `delete()` method, thanks! :)